Fe surface distance evaluator

drwells opened the pull request:
Another spinoff from #418. This class is needed in a few different places.

I think this is the last (or nearly last) patch we can pull from the wave tank branch easily.

amneetb commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+     *                           <------->
+     *                            box_half_dx[0]      vert1         vert2
+     */
+    static bool checkIntersection3D(const IBTK::Vector3d& box_center,
```
@drwells : I hope you liked my ASCII art.

drwells replied:
It makes the point clearly, so yes :)

boyceg commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+class FESurfaceDistanceEvaluator
+{
+public:
+    static const double s_large_distance;
```
If this is meant to be used outside of the class, let's stick with the standard convention:
```suggestion
    static const double LARGE_DISTANCE;
```

boyceg commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+     */
+    FESurfaceDistanceEvaluator(const std::string& object_name,
+                               SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > patch_hierarchy,
+                               SAMRAI::tbox::Pointer<IBAMR::IBFEMethod> ibfe_method,
```
Does this need the full `IBFEMethod` implementation, or is it good enough to have the `FEDataManager` or even just the `Mesh`?

boyceg replied:
Looking at the implementation, it appears that you could _possibly_ just pass in the `EquationSystems` object and omit `ibfe_method`, `mesh`, and `bdry_mesh`.

boyceg commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+                               SAMRAI::tbox::Pointer<SAMRAI::hier::PatchHierarchy<NDIM> > patch_hierarchy,
+                               SAMRAI::tbox::Pointer<IBAMR::IBFEMethod> ibfe_method,
+                               const libMesh::Mesh& mesh,
+                               const libMesh::BoundaryMesh& bdry_mesh,
```
Do we need to have `BoundaryMesh` in the interface here --- can we just get the boundary mesh internally, or does the `BoundaryMesh` need to be properly set up for this to work?

boyceg commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+                               SAMRAI::tbox::Pointer<IBAMR::IBFEMethod> ibfe_method,
+                               const libMesh::Mesh& mesh,
+                               const libMesh::BoundaryMesh& bdry_mesh,
+                               const int part,
```
If you have the part number and you have to pass in the `IBFEMethod` implementation, can we extract the `Mesh` internally?

boyceg commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+    ~FESurfaceDistanceEvaluator();
+
+    /*!
+     * Map the triangles intersecting a particular grid cell.
```
```suggestion
     * Map the triangles intersecting a particular Cartesian grid cell.
```
Also, is this specialized to triangular meshes? If not, suggest:
```suggestion
     * Map the elements from the finite element mesh intersecting a particular Cartesian grid cell.
```

boyceg commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+    /*!
+     * Destructor for this class.
+     */
+    ~FESurfaceDistanceEvaluator();
```
Probably can do:
```suggestion
    ~FESurfaceDistanceEvaluator() = default;
```

boyceg commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+    /*!
+     * Get the map maintaining triangle-cell intersection and neighbors.
+     */
+    const std::map<SAMRAI::pdat::CellIndex<NDIM>, std::set<libMesh::Elem*>, IBTK::CellIndexFortranOrder>&
```
```suggestion
    using intersection_map = std::map<SAMRAI::pdat::CellIndex<NDIM>, std::set<libMesh::Elem*>, IBTK::CellIndexFortranOrder>;
    const intersection_map&
```

boyceg commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+    /*!
+     * Compute the signed distance in the viscinity of the finite element mesh.
+     */
+    void computeSignedDistance(int n_idx, int d_idx);
```
What are `n_idx` and `d_idx`?

boyceg commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+    void computeSignedDistance(int n_idx, int d_idx);
+
+    /*!
+     * Update the sign of the distance function away from the finite element mesh.
```
It might be helpful to briefly describe how this is different from `computeSignedDistance()`, and how they are meant to be used.

amneetb commented on src/level_set/FESurfaceDistanceEvaluator.cpp:
```diff
@@ -0,0 +1,1310 @@
+                            for (int d = 0; d < NDIM; ++d)
+                            {
+                                avg_unit_normal(d) += bdry_normals[qp](d);
+                            }
```
Since this class is of interest, let's use the normal in the right way. We should follow the angle weighted normal as explained here:
https://ieeexplore.ieee.org/document/1407857
https://www.sciencedirect.com/science/article/pii/S0021999114004495

@drwells : Would like to implement this?

boyceg commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+     *                |              |
+     *     (box_bl)   *--------------*  (box_br)
+     */
+    static bool checkIntersection2D(const IBTK::Vector3d& box_bl,
```
It might make sense to put these in `libmesh_utilities.h`?

Also, @drwells, could we use the box class here that you are using for the new partitioner?

boyceg commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+    /*!
+     * Default assignment operator is not implemented and should not be used.
+     */
+    FESurfaceDistanceEvaluator& operator=(const FESurfaceDistanceEvaluator& that);
```
```suggestion
    FESurfaceDistanceEvaluator& operator=(const FESurfaceDistanceEvaluator&) = delete;
```

boyceg commented on include/ibamr/FESurfaceDistanceEvaluator.h:
```diff
@@ -0,0 +1,249 @@
+    /*!
+     * Default copy constructor is not implemented and should not be used.
+     */
+    FESurfaceDistanceEvaluator(const FESurfaceDistanceEvaluator& from);
```
```suggestion
    FESurfaceDistanceEvaluator(const FESurfaceDistanceEvaluator&) = delete;
```

amneetb commented:
It will be efficient (to avoid back-and-forth) if you guys can make these changes according to the design you have in mind. I approve all of the above changes. Some clarifications are below:

`n_idx` is the patch data index to plot the number of elements per Cartesian Cell in VisIt and 'd_idx` is the patch data index to hold the actual distance function.  

LARGE_DISTANCE should be set by the user according to their application. For example if distance is calculated for only 4 cells near the interface, then I generally set LARGE DISTANCE to 5*h. But for some applications, LARGE DISTANCE can actually be a very large value.

amneetb commented:
What I had in my mind with all the design decisions I had chosen are already there. I can't think of anything better, but I am open to and welcome new changes.

boyceg commented:
> What I had in my mind with all the design decisions I had chosen are already there. I can't think of anything better, but I am open to and welcome new changes.

The design is fine.

My questions were trying to get at whether this thing really needed any state or not --- like could it just be a collection of static functions? --- and if it needed state, how much needed to be provided.

If you can update the comments to address the questions about uses/limitations of the class, I will try to strip out unneeded stuff in the interface, and then we can get this merged.

amneetb commented:
Done.

In case you want to see the current usage of the class:
https://github.com/IBAMR/IBAMR/blob/amneetb/numerical_wave_tank/examples/IBLevelSet/ex0/example.cpp#L829-L868

drwells commented:
We discussed this by email but I don't think we reached a conclusion regarding who is going to finish addressing all of the review comments, add tests, etc. Should we explicitly assign this to a student or to someone else?

drwells commented:
One thing that we should add is (from a conversation with @amneetb) the use of angle-weighted normals in the case where multiple faces/edges are at an equal distance from the point: see

"An efficient immersed boundary treatment for complex moving object" by Chen Liu and Changhong Hu (JCP, 2014)

drwells commented:
We should try to make sure that this is still second order accurate even if we have higher degree finite elements and isoparametric elements.

drwells commented:
We should add a reference to the relevant paper that describes the algorithm used here.

drwells commented:
Superceded by #744.
